Check volume's status before detaching volume

FengyunPan opened the pull request:
When volume's status is 'detaching', controllermanager will detach
it again and return err. It is necessary to check volume's status
before detaching volume.

same issue: #44536

CaoShuFeng commented on pkg/volume/cinder/attacher.go:
```diff
@@ -278,6 +278,15 @@ func (detacher *cinderDiskDetacher) Detach(deviceMountPath string, nodeName type
 
+	pending, volumeStatus, err := detacher.cinderProvider.OperationPending(volumeID)
+	if err != nil {
+		// Continue attach, the same as DiskIsAttached()
```
Continue detach?

FengyunPan replied:
Yes, the same as DiskIsAttached().
If OperationPending() return err, DiskIsAttached() will return same err, we can ignore it and try detach anyway.

CaoShuFeng replied:
I mean replace "attach" with "detach" in this line.
：）

FengyunPan replied:
Niu Bi !!!

FengyunPan commented:
/release-note-none

msau42 commented on pkg/volume/cinder/attacher.go:
```diff
@@ -75,9 +78,9 @@ func (attacher *cinderDiskAttacher) Attach(spec *volume.Spec, nodeName types.Nod
 
 	pending, volumeStatus, err := attacher.cinderProvider.OperationPending(volumeID)
```
Why does OperationPending return error if it is pending?  Shouldn't err == nil and pending == true?

FengyunPan replied:
Thank you very much, your comments are very helpful to me. I update it soon.

FengyunPan replied:
When OperationPending() failed to get volume, it return error and pending=false.
When OperationPending() get volume and volume's status is pending, it return pending=true/err=nil.
Or return pending=false/err=nil.

msau42 commented on pkg/volume/cinder/attacher.go:
```diff
@@ -75,7 +75,7 @@ func (attacher *cinderDiskAttacher) Attach(spec *volume.Spec, nodeName types.Nod
 	pending, volumeStatus, err := attacher.cinderProvider.OperationPending(volumeID)
 	if err != nil {
-		return "", err
+		// Continue attach, the same as DiskIsAttached()
```
When does OperationPending fail?  Do you want to continue even if it fails? What could happen if two attach or two detach operations occur at the same time?  Or an attach and a detach?

FengyunPan replied:
I see, I should return err when OperationPending() or DiskIsAttached() return err.
The err will be 'Unable to initialize cinder client', return err better.

FengyunPan commented:
/assign @jsafrane

msau42 commented on pkg/volume/cinder/attacher.go:
```diff
@@ -88,6 +88,7 @@ func (attacher *cinderDiskAttacher) Attach(spec *volume.Spec, nodeName types.Nod
 		glog.Warningf(
 			"Error checking if volume (%q) is already attached to current instance (%q). Will continue and try attach anyway. err=%v",
 			volumeID, instanceid, err)
+		return "", err
```
@jsafrane Are you familiar with the implications of continuing to attach here?  Should we keep the old behavior or change it to return an error?

jsafrane commented on pkg/volume/cinder/attacher.go:
```diff
@@ -278,12 +279,22 @@ func (detacher *cinderDiskDetacher) Detach(deviceMountPath string, nodeName type
+	}
+	if pending {
+		glog.Infof("status of volume %q is %s, wait it to be finished", volumeID, volumeStatus)
+		return volume.ErrNotFinish
```
I think that you should wait for the operation to finish here (with some sane timeout). You should not mis-use attach/detach controller loop to ignore some errors

FengyunPan replied:
Yeah, I will do it.
By the way, should we keep the old behavior or change it to return an error at Line 143 / 91 / 297 ?

msau42 commented:
@k8s-bot ok to test

msau42 commented on pkg/volume/cinder/attacher.go:
```diff
@@ -293,10 +329,14 @@ func (detacher *cinderDiskDetacher) Detach(deviceMountPath string, nodeName type
 	}
 
 	if err = detacher.cinderProvider.DetachDisk(instanceID, volumeID); err != nil {
```
So if we return timeout error, then the detach controller will retry again later.  Do we still need to check if operation is pending?  Or can the cinder provider handle duplicate detach calls?

FengyunPan replied:
Yes,  the cinder provider can't handle duplicate detach, we need to check if operation is pending. I update it soon.

FengyunPan commented:
@k8s-bot pull-kubernetes-verify test this
@k8s-bot pull-kubernetes-e2e-kops-aws test this
@k8s-bot pull-kubernetes-unit test this

FengyunPan commented:
@k8s-bot pull-kubernetes-verify test this

jsafrane commented:
@FengyunPan, verify tests are very rarely wrong and they inform you very clearly what needs to be done:

```
Verifying hack/make-rules/../../hack/verify-bazel.sh
...
W0529 20:24:29.310] 1 BUILD files not up-to-date.
I0529 20:24:29.411] 
I0529 20:24:29.411] Run ./hack/update-bazel.sh
I0529 20:24:29.411] FAILED   hack/make-rules/../../hack/verify-bazel.sh	6s
...

Please run hack/update-bazel.sh and that's it.

jsafrane commented on pkg/volume/cinder/attacher.go:
```diff
@@ -278,6 +314,15 @@ func (detacher *cinderDiskDetacher) Detach(deviceMountPath string, nodeName type
+	}
+	if pending {
+		glog.Infof("status of volume %q is %s, wait it to be finished", volumeID, volumeStatus)
+		return nil
```
If you return `nil` here then it means that the volume is detached, right? You should probably wait here for the pending operation to complete (with the same exponential backoff as above?)

FengyunPan replied:
Yes, you are right, I update it soon.

msau42 commented on pkg/volume/cinder/attacher.go:
```diff
@@ -267,6 +274,63 @@ func (plugin *cinderPlugin) NewDetacher() (volume.Detacher, error) {
+	})
+
+	if err == wait.ErrWaitTimeout {
+		err = fmt.Errorf("Volume %q is %s, cann't finish within the alloted time", volumeID, volumeStatus)
```
s/cann't/can't

FengyunPan replied:
Done

msau42 commented on pkg/volume/cinder/attacher.go:
```diff
@@ -293,10 +361,14 @@ func (detacher *cinderDiskDetacher) Detach(deviceMountPath string, nodeName type
+		return err
+	}
+	if err = detacher.waitDiskDetached(instanceID, volumeID); err != nil {
+		glog.Errorf("Error detaching volume %q from node %q: %v", volumeID, nodeName, err)
```
A better error msg could be "Error waiting for volume %q to detach from node %q: %v"

FengyunPan replied:
Nice, thank you very much.

Commit:
Wait for detach operation to complete

When volume's status is 'detaching', controllermanager will detach
it again and return err. It is necessary to wait for detach
operation to complete within the alloted time.

msau42 commented:
/lgtm

msau42 commented:
@k8s-bot pull-kubernetes-kubemark-e2e-gce test this

jsafrane commented:
/approve
